flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArraySliceFunction.java

```diff
@@ -0,0 +1,121 @@
+import org.apache.flink.table.functions.BuiltInFunctionDefinitions;
+import org.apache.flink.table.functions.SpecializedFunction;
+import org.apache.flink.table.types.CollectionDataType;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.util.FlinkRuntimeException;
+
+import javax.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/** Implementation of {@link BuiltInFunctionDefinitions#ARRAY_SLICE}. */
+@Internal
+public class ArraySliceFunction extends BuiltInScalarFunction {
+    private final ArrayData.ElementGetter elementGetter;
+
+    public ArraySliceFunction(SpecializedFunction.SpecializedContext context) {
+        super(BuiltInFunctionDefinitions.ARRAY_SLICE, context);
+        final DataType dataType =
+                ((CollectionDataType) context.getCallContext().getArgumentDataTypes().get(0))
+                        .getElementDataType();
+        elementGetter = ArrayData.createElementGetter(dataType.getLogicalType());
+    }
+
+    public @Nullable ArrayData eval(ArrayData array, Integer start, Integer end) {
+        try {
+            if (array == null || start == null || end == null) {
+                return null;
+            }
+            if (array.size() == 0) {
+                return array;
+            }
+
+            Integer startIndex = null;
+            Integer endIndex = null;
+
+            if (start == 0) {
+                start = 1;
+            }
+            if (start < 0) {
+                if (start < -array.size()) {
+                    startIndex = start;
+                } else {
+                    startIndex = array.size() + start;
+                }
+            }
+            if (start > 0) {
+                if (start > array.size()) {
+                    startIndex = start;
+                } else {
+                    startIndex = start - 1;
+                }
+            }
+            if (end == 0) {
+                end = 1;
+            }
+            if (end < 0) {
+                if (end < -array.size()) {
+                    endIndex = end;
+                } else {
+                    endIndex = array.size() + end;
+                }
+            }
+            if (end > 0) {
+                if (end > array.size()) {
+                    endIndex = end;
+                } else {
+                    endIndex = end - 1;
+                }
+            }
+
+            if (startIndex > endIndex) {
+                return new GenericArrayData(new Object[0]);
+            }
+            if (startIndex > array.size()) {
+                return new GenericArrayData(new Object[0]);
+            }
+            if (endIndex > array.size()) {
+                endIndex = array.size() - 1;
+            }
+            if (startIndex < -array.size()) {
+                startIndex = 0;
+            }
+            if (endIndex < -array.size()) {
+                return new GenericArrayData(new Object[0]);
+            }
```

Comment:
If this code block doesn't exit logic problem. I will concise it at next commit.